File: ios/Sources/UWBCompassCore/SessionStateMachine.swift
import Foundation

/// Client-side mirror of the backend session state machine (ADR-0007).
public enum SessionState: Sendable {
    case idle, invited, incoming, accepted, negotiated, active, ended, declined, expired, failed
}

public struct IllegalTransition: Error { public let from: SessionState; public let to: SessionState }

/// Guards client-side session transitions against the legal graph.
public final class SessionStateMachine {
    public private(set) var state: SessionState

    public init(initial: SessionState = .idle) { self.state = initial }

    private static let allowed: [String: Set<String>] = [
        key(.idle): [key(.invited), key(.incoming)],
        key(.invited): [key(.accepted), key(.declined), key(.expired), key(.failed), key(.ended)],
        key(.incoming): [key(.accepted), key(.declined), key(.expired), key(.failed), key(.ended)],
        key(.accepted): [key(.negotiated), key(.expired), key(.failed), key(.ended)],
        key(.negotiated): [key(.active), key(.expired), key(.failed), key(.ended)],
        key(.active): [key(.ended), key(.failed)],
        key(.ended): [], key(.declined): [], key(.expired): [], key(.failed): [],
    ]

    public var isTerminal: Bool {
        switch state {
        case .ended, .declined, .expired, .failed: return true
        default: return false
        }
    }

    public func canTransition(to: SessionState) -> Bool {
        (Self.allowed[Self.key(state)] ?? []).contains(Self.key(to))
    }

    @discardableResult
    public func transition(to: SessionState) throws -> SessionState {
        guard canTransition(to: to) else { throw IllegalTransition(from: state, to: to) }
        state = to
        return state
    }

    /// Returns to `.idle` from any terminal state (back to the peer list).
    public func resetIfTerminal() {
        if isTerminal { state = .idle }
    }

    private static func key(_ s: SessionState) -> String { String(describing: s) }
}

File: android/core/src/main/kotlin/com/uwbcompass/core/SessionStateMachine.kt
package com.uwbcompass.core

/** Client-side mirror of the backend session state machine (ADR-0007). */
enum class SessionState { IDLE, INVITED, INCOMING, ACCEPTED, NEGOTIATED, ACTIVE, ENDED, DECLINED, EXPIRED, FAILED }

/**
 * Reduces server events into the client's local session state and guards against
 * illegal transitions. `IDLE` = no session; `INVITED` = we invited someone (outgoing);
 * `INCOMING` = someone invited us. Both funnel into ACCEPTED/NEGOTIATED/ACTIVE.
 */
class SessionStateMachine(initial: SessionState = SessionState.IDLE) {
    var state: SessionState = initial
        private set

    private val allowed: Map<SessionState, Set<SessionState>> = mapOf(
        SessionState.IDLE to setOf(SessionState.INVITED, SessionState.INCOMING),
        SessionState.INVITED to setOf(SessionState.ACCEPTED, SessionState.DECLINED, SessionState.EXPIRED, SessionState.FAILED, SessionState.ENDED),
        SessionState.INCOMING to setOf(SessionState.ACCEPTED, SessionState.DECLINED, SessionState.EXPIRED, SessionState.FAILED, SessionState.ENDED),
        SessionState.ACCEPTED to setOf(SessionState.NEGOTIATED, SessionState.EXPIRED, SessionState.FAILED, SessionState.ENDED),
        SessionState.NEGOTIATED to setOf(SessionState.ACTIVE, SessionState.EXPIRED, SessionState.FAILED, SessionState.ENDED),
        SessionState.ACTIVE to setOf(SessionState.ENDED, SessionState.FAILED),
        SessionState.ENDED to emptySet(),
        SessionState.DECLINED to emptySet(),
        SessionState.EXPIRED to emptySet(),
        SessionState.FAILED to emptySet(),
    )

    val isTerminal: Boolean
        get() = state in setOf(SessionState.ENDED, SessionState.DECLINED, SessionState.EXPIRED, SessionState.FAILED)

    fun canTransition(to: SessionState): Boolean = to in (allowed[state] ?: emptySet())

    /** @throws IllegalStateException on an illegal transition. */
    fun transition(to: SessionState): SessionState {
        check(canTransition(to)) { "illegal client transition $state -> $to" }
        state = to
        return state
    }

    /** Resets to IDLE from any terminal state (returning to the peer list). */
    fun resetIfTerminal() {
        if (isTerminal) state = SessionState.IDLE
    }
}
